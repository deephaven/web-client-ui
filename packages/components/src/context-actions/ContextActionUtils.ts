import { IconDefinition } from '@deephaven/icons';
import React from 'react';
import type { Shortcut } from '../shortcuts';

export type ResolvableContextAction =
  | ContextAction
  | Promise<ContextAction[]>
  | (() => Promise<ContextAction[]> | ContextAction[] | ContextAction);

export type MenuItem = ContextAction | Promise<ContextAction[]>;

export interface ContextAction {
  title?: string;
  description?: string;
  action?(event?: KeyboardEvent): void;
  actions?: ResolvableContextAction[];
  icon?: IconDefinition | React.ReactElement;
  iconColor?: string;
  shortcut?: Shortcut;
  isGlobal?: boolean;
  group?: number;
  order?: number;
  disabled?: boolean;
  menuElement?: React.ReactElement;
  iconOutline?: boolean;
}

export interface ContextActionEvent extends MouseEvent {
  contextActions: ResolvableContextAction[];
}

class ContextActionUtils {
  static actionsDisabled = false;

  static disableAllActions(): void {
    ContextActionUtils.actionsDisabled = true;
  }

  static enableAllActions(): void {
    ContextActionUtils.actionsDisabled = false;
  }

  static isContextActionEvent(e: MouseEvent): e is ContextActionEvent {
    return Array.isArray((e as ContextActionEvent).contextActions);
  }

  /**
   * Compare two action items. Useful in Array.sort
   * @param a First context action to compare
   * @param b Second context action to compare
   */
  static compareActions(a: ContextAction, b: ContextAction): number {
    if (a.group !== b.group) {
      return (a.group || 0) > (b.group || 0) ? 1 : -1;
    }

    if (a.order !== b.order) {
      return (a.order || 0) > (b.order || 0) ? 1 : -1;
    }

    if (a.title !== b.title) {
      return (a.title || '') > (b.title || '') ? 1 : -1;
    }

    if (a !== b) {
      return a > b ? 1 : -1;
    }

    return 0;
  }

  /**
   *
   * @param actions The array of actions to sort
   */
  static sortActions(actions: ContextAction[]): ContextAction[] {
    if (!actions || !Array.isArray(actions)) {
      return [];
    }

    const sortedActions = actions.slice();
    sortedActions.sort(ContextActionUtils.compareActions);
    return sortedActions;
  }

  static isMacPlatform(): boolean {
    const { platform } = window.navigator;
    return platform.startsWith('Mac');
  }

  /**
   * Retrieve the preferred modifier key based on the current platform
   */
  static getModifierKey(): 'metaKey' | 'ctrlKey' {
    if (ContextActionUtils.isMacPlatform()) {
      return 'metaKey';
    }

    return 'ctrlKey';
  }

  /**
   * Returns true if the modifier key for the current platform is down for the event (Ctrl for windows/linux, Command (meta) for mac)
   * @param event The event to get the meta key status from
   */
  static isModifierKeyDown(
    event: KeyboardEvent | MouseEvent | React.KeyboardEvent | React.MouseEvent
  ): boolean {
    const modifierKey = ContextActionUtils.getModifierKey();
    return event[modifierKey];
  }

  /**
   * Copy the passed in text to the clipboard.
   * @param text The text to copy
   * @returns Promise Resolved on success, rejected on failure
   */
  static async copyToClipboard(text: string): Promise<void> {
    try {
      const permissions = await navigator.permissions.query({
        name: 'clipboard-write' as PermissionName,
      });
      if (permissions.state !== 'granted' && permissions.state !== 'prompt') {
        throw new Error('Invalid permissions for clipboard-write');
      }
    } catch (e) {
      // Fallback if we can't get permissions for some reason
      ContextActionUtils.copyToClipboardExecCommand(text);
      return;
    }

    await navigator.clipboard.writeText(text);
  }

  /**
   * Copy the passed in text to the clipboard using the `execCommand` functionality
   * Throws on error/failure
   * @param text The text to copy
   */
  static copyToClipboardExecCommand(text: string): void {
    const oldFocus = document.activeElement;
    const textArea = document.createElement('textarea');
    textArea.value = text;
    document.body.appendChild(textArea);
    textArea.focus();
    textArea.select();

    if (!document.execCommand('copy')) {
      throw new Error('Unable to execute copy command');
    }

    document.body.removeChild(textArea);

    if (oldFocus instanceof HTMLElement) {
      oldFocus.focus();
    }
  }

  /**
   * Returns the menu items for the provided context actions, or empty array if none found.
   * @param actionsParam The actions to get menu items for
   * @param includePromises Whether or not to include promises in the returned menu items
   */
  static getMenuItems(
    actionsParam: ResolvableContextAction | ResolvableContextAction[],
    includePromises?: true
  ): MenuItem[];

  // If ignoring promises, then the return type is narrowed
  static getMenuItems(
    actionsParam: ResolvableContextAction | ResolvableContextAction[],
    includePromises: false
  ): ContextAction[];

  static getMenuItems(
    actionsParam: ResolvableContextAction | ResolvableContextAction[],
    includePromises = true
  ): MenuItem[] {
    let menuItems: MenuItem[] = [];
    let actions = actionsParam;
    if (!Array.isArray(actions)) {
      actions = [actions];
    }

    for (let i = 0; i < actions.length; i += 1) {
      const action = actions[i];
      let newMenuItems:
        | ContextAction
        | ContextAction[]
        | Promise<ContextAction[]>;
      if (typeof action === 'function') {
        newMenuItems = action();
      } else {
        newMenuItems = action;
      }

      if (newMenuItems != null) {
        if (newMenuItems instanceof Promise) {
          if (includePromises) {
            menuItems.push(newMenuItems);
          }
        } else if (Array.isArray(newMenuItems)) {
          menuItems = menuItems.concat(newMenuItems);
        } else {
          menuItems.push(newMenuItems);
        }
      }
    }

    menuItems = menuItems.filter(
      action =>
        (action as ContextAction).title ||
        (action as Promise<ContextAction[]>).then ||
        (action as ContextAction).menuElement
    );

    return menuItems;
  }

  /**
   * Returns the index of the next menu item in a list that doesn't have a disabled=true prop
   * @param startIndex the starting position for the iteration
   * @param delta the direction of travel, -1 or 1
   * @param menuItems an array of menuItems
   */
  static getNextMenuItem(
    startIndex: number,
    delta: -1 | 1,
    menuItems: MenuItem[]
  ): number {
    let firstIndex = startIndex;
    if (firstIndex < 0 && delta < 0) {
      // if menu index is -1 and delta -1 manually set start point
      firstIndex = menuItems.length;
    }
    // find the next non disabled menu option, iterating the list only once
    for (let i = 1; i < menuItems.length + 1; i += 1) {
      const menuIndex =
        (firstIndex + delta * i + menuItems.length) % menuItems.length;
      const item = menuItems[menuIndex];
      if (!(item instanceof Promise) && item.disabled !== true) {
        return menuIndex;
      }
    }
    return startIndex;
  }
}

export default ContextActionUtils;
