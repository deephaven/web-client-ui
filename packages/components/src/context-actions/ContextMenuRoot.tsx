import React, { Component } from 'react';
import classNames from 'classnames';
import ContextMenu from './ContextMenu';
import ContextActionUtils, { MenuItem } from './ContextActionUtils';

type ContextMenuRootProps = Record<string, never> & {
  'data-testid'?: string;
};

interface ContextMenuRootState {
  actions: MenuItem[] | null;
  left: number;
  top: number;
}

/**
 * Put at your root container, any contextmenu events that are unhandled in the root container will be handled by this
 */
class ContextMenuRoot extends Component<
  ContextMenuRootProps,
  ContextMenuRootState
> {
  constructor(props: ContextMenuRootProps) {
    super(props);

    this.handleMenuClose = this.handleMenuClose.bind(this);
    this.handleContextMenu = this.handleContextMenu.bind(this);
    this.container = React.createRef();
    this.openMenu = React.createRef();

    this.state = {
      actions: null,
      left: 0,
      top: 0,
    };
  }

  componentDidMount(): void {
    if (this.container.current?.parentElement) {
      this.container.current.parentElement.addEventListener(
        'contextmenu',
        this.handleContextMenu
      );
    }
  }

  componentWillUnmount(): void {
    if (this.container.current?.parentElement) {
      this.container.current.parentElement.removeEventListener(
        'contextmenu',
        this.handleContextMenu
      );
    }
  }

  container: React.RefObject<HTMLDivElement>;

  openMenu: React.RefObject<ContextMenu>;

  handleContextMenu(e: MouseEvent): void {
    if (!this.container.current) {
      return;
    }

    const parentRect = this.container.current.getBoundingClientRect();
    const top = e.clientY - parentRect.top;
    const left = e.clientX - parentRect.left;
    const { actions } = this.state;

    // Context menu is open and user clicked on the context-root blocking layer
    // Mac and Linux appear to trigger contextmenu events on mousedown vs. mouseup on Windows
    // Mouseup on Windows triggers blur before contextmenu which effectively does what this path does
    if (actions != null && e.target === this.container.current) {
      // re-emit right clicks that hit the context-root blocking layer
      // while we already have a custom context menu open
      e.preventDefault();

      // Set actions to null removes the menu
      // That allows a new menu to be opened on a different element so initial position is set properly
      // Otherwise the instance of this menu may be reused
      // A new contextmenu event is triggered on the element at the location the user clicked on the blocking layer
      this.setState({ actions: null }, () => {
        const element = document.elementFromPoint(left, top); // x y

        const mouseEvent = new MouseEvent('contextmenu', {
          clientX: e.clientX,
          clientY: e.clientY,
          bubbles: true,
          cancelable: true,
        });

        element?.dispatchEvent(mouseEvent);
      });
      return;
    }

    if (!ContextActionUtils.isContextActionEvent(e)) {
      // Open native menu if no custom context actions
      return;
    }

    if (e.metaKey || e.ctrlKey) {
      // debug escape hatch to native menu
      return;
    }

    const contextActions = ContextActionUtils.getMenuItems(e.contextActions);

    if (contextActions.length === 0) {
      // No actions after filtering. Use native menu
      return;
    }

    // new clicks, set actions
    e.preventDefault();
    this.setState({
      actions: contextActions,
      top,
      left,
    });
  }

  handleMenuClose(menu: ContextMenu): void {
    if (menu === this.openMenu.current) {
      this.setState({ actions: null });
    }
  }

  render(): JSX.Element {
    let menu = null;
    const { 'data-testid': dataTestId } = this.props;
    const { actions, top, left } = this.state;
    if (actions) {
      menu = (
        <ContextMenu
          ref={this.openMenu}
          actions={actions}
          onMenuClosed={this.handleMenuClose}
          top={top}
          left={left}
          updatePosition={(verifiedTop, verifiedLeft) => {
            this.setState({ top: verifiedTop, left: verifiedLeft });
          }}
          data-testid={dataTestId}
        />
      );
    }
    return (
      <div
        className={classNames('context-menu-root', { active: actions })}
        ref={this.container}
      >
        {menu}
      </div>
    );
  }
}

export default ContextMenuRoot;
