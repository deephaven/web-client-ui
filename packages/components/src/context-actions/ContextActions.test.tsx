import React from 'react';
import TestRenderer from 'react-test-renderer';
import { waitFor } from '@testing-library/react';
import ContextActionUtils from './ContextActionUtils';
import ContextMenuRoot from './ContextMenuRoot';

let spy: jest.SpyInstance;
beforeEach(() => {
  spy = jest.spyOn(window, 'requestAnimationFrame').mockImplementation(cb => {
    // mock request animation frame
    // see https://github.com/deephaven/web-client-ui/issues/508
    // only safe to mock like this if RAF is non-recursive
    cb(0);
    return 0;
  });
});

afterEach(() => {
  spy.mockRestore();
});

type ContextMenuMock = {
  addEventListener: jest.Mock<void>;
  removeEventListener: jest.Mock<void>;
  getBoundingClientRect: jest.Mock<{
    top: number;
    left: number;
    width: number;
    height: number;
  }>;
  focus: jest.Mock<void>;
  setAttribute: jest.Mock<void>;
  offsetParent: ContextMenuMock | null;
  parentElement: ContextMenuMock | null;
};

const TEST_MENU_1 = [
  { title: 'Test1' },
  { title: 'Test2' },
  { title: 'Test3' },
];

function contextMenuMock(
  mock: ContextMenuMock,
  element: JSX.Element
): ContextMenuMock | null {
  if (element.props != null && element.props.className != null) {
    if (element.props.className.indexOf('context-menu') !== -1) {
      return mock;
    }
  }

  return null;
}

function DEFAULT_MOCK(mockParent = true): ContextMenuMock {
  return {
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    getBoundingClientRect: jest.fn(() => ({
      top: 0,
      left: 0,
      width: 0,
      height: 0,
    })),
    focus: jest.fn(),
    offsetParent: mockParent ? DEFAULT_MOCK(false) : null,
    parentElement: mockParent ? DEFAULT_MOCK(false) : null,
    setAttribute: jest.fn(),
  };
}

it('renders without crashing', () => {
  const mock = DEFAULT_MOCK();

  const tree = TestRenderer.create(
    <div>
      <ContextMenuRoot />
    </div>,
    {
      createNodeMock: contextMenuMock.bind(this, mock),
    }
  );

  expect(mock.parentElement?.addEventListener.mock.calls.length).toBe(1);
  expect(mock.parentElement?.addEventListener.mock.calls[0][0]).toBe(
    'contextmenu'
  );
  expect(mock.parentElement?.addEventListener.mock.calls[0][1]).not.toBeNull();
  expect(mock.parentElement?.removeEventListener.mock.calls.length).toBe(0);

  expect(tree).toMatchSnapshot();

  tree.unmount();

  expect(mock.parentElement?.addEventListener.mock.calls.length).toBe(1);
  expect(mock.parentElement?.removeEventListener.mock.calls.length).toBe(1);
});

it('handles the context menu event on the parent, renders an array of menu items', () => {
  const mock = DEFAULT_MOCK();

  const tree = TestRenderer.create(
    <div>
      <ContextMenuRoot />
    </div>,
    {
      createNodeMock: contextMenuMock.bind(this, mock),
    }
  );

  expect(mock.parentElement?.addEventListener.mock.calls.length).toBe(1);
  expect(mock.parentElement?.addEventListener.mock.calls[0][0]).toBe(
    'contextmenu'
  );
  expect(mock.parentElement?.addEventListener.mock.calls[0][1]).not.toBeNull();
  expect(mock.parentElement?.removeEventListener.mock.calls.length).toBe(0);

  expect(tree.root.findAllByType('button').length).toBe(0);

  expect(tree).toMatchSnapshot();

  const mockEvent = { preventDefault: jest.fn(), contextActions: TEST_MENU_1 };
  const handleContextMenu =
    mock.parentElement?.addEventListener.mock.calls[0][1];
  handleContextMenu(mockEvent);

  expect(tree.root.findAllByType('button').length).toBe(TEST_MENU_1.length);

  tree.unmount();

  expect(mock.parentElement?.addEventListener.mock.calls.length).toBe(1);
  expect(mock.parentElement?.removeEventListener.mock.calls.length).toBe(1);
});

it('renders a promise returning menu items properly', async () => {
  const mock = DEFAULT_MOCK();
  const promise = Promise.resolve(TEST_MENU_1);

  const tree = TestRenderer.create(
    <div>
      <ContextMenuRoot />
    </div>,
    {
      createNodeMock: contextMenuMock.bind(this, mock),
    }
  );

  const mockEvent = { preventDefault: jest.fn(), contextActions: [promise] };
  const handleContextMenu =
    mock.parentElement?.addEventListener.mock.calls[0][1];
  handleContextMenu(mockEvent);

  await waitFor(() =>
    expect(tree.root.findAllByType('button').length).toBe(TEST_MENU_1.length)
  );
  expect(tree).toMatchSnapshot();
});

it('renders an empty menu for a rejected promise', () => {
  const mock = DEFAULT_MOCK();
  const promise = Promise.reject();

  const tree = TestRenderer.create(
    <div>
      <ContextMenuRoot />
    </div>,
    {
      createNodeMock: contextMenuMock.bind(this, mock),
    }
  );

  const mockEvent = { preventDefault: jest.fn(), contextActions: [promise] };
  const handleContextMenu =
    mock.parentElement?.addEventListener.mock.calls[0][1];
  handleContextMenu(mockEvent);

  expect(tree.root.findAllByType('button').length).toBe(0);
  expect(tree).toMatchSnapshot();
});

it('renders a menu from a promise returned from a function', () => {
  const mock = DEFAULT_MOCK();
  const promise = Promise.resolve([]);
  const fn = () => promise;

  const tree = TestRenderer.create(
    <div>
      <ContextMenuRoot />
    </div>,
    {
      createNodeMock: contextMenuMock.bind(this, mock),
    }
  );

  const mockEvent = { preventDefault: jest.fn(), contextActions: [fn] };
  const handleContextMenu =
    mock.parentElement?.addEventListener.mock.calls[0][1];
  handleContextMenu(mockEvent);

  expect(tree.root.findAllByType('button').length).toBe(0);
  expect(tree).toMatchSnapshot();
});

it('sorts by title properly', () => {
  const actions = [{ title: 'b' }, { title: 'c' }, { title: 'a' }];
  const expectedResult = [actions[2], actions[0], actions[1]];
  const sortedActions = ContextActionUtils.sortActions(actions);
  expect(sortedActions).toEqual(expectedResult);
});

it('sorts by order properly', () => {
  const actions = [
    { title: 'c', order: 40 },
    { title: 'b', order: 90 },
    { title: 'a', order: 10 },
  ];
  const expectedResult = [actions[2], actions[0], actions[1]];
  const sortedActions = ContextActionUtils.sortActions(actions);
  expect(sortedActions).toEqual(expectedResult);
});

it('sorts groups properly', () => {
  const actions = [
    { title: 'c', order: 40, group: 10 },
    { title: 'b', order: 90, group: 10 },
    { title: 'a', order: 10, group: 30 },
  ];
  const expectedResult = [actions[0], actions[1], actions[2]];
  const sortedActions = ContextActionUtils.sortActions(actions);
  expect(sortedActions).toEqual(expectedResult);
});

it('sorts mixed groups properly', () => {
  const actions = [
    { title: 'c', order: 40, group: 10 },
    { title: 'a', order: 10 },
    { title: 'b', order: 90, group: 10 },
  ];
  const expectedResult = [actions[1], actions[0], actions[2]];
  const sortedActions = ContextActionUtils.sortActions(actions);
  expect(sortedActions).toEqual(expectedResult);
});

it('sorts groups/orders properly', () => {
  const actions = [
    { title: 'c', order: 104, group: 1 },
    { order: 10, group: 1 },
    { title: 'b', order: 105, group: 1 },
  ];
  const expectedResult = [actions[1], actions[0], actions[2]];
  const sortedActions = ContextActionUtils.sortActions(actions);
  expect(sortedActions).toEqual(expectedResult);
});
