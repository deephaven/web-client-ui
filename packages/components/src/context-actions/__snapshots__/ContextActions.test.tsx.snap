// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles the context menu event on the parent, renders an array of menu items 1`] = `
<div>
  <div
    className="context-menu-root"
  />
</div>
`;

exports[`renders a menu from a promise returned from a function 1`] = `
<div>
  <div
    className="context-menu-root active"
  >
    <div
      className="context-menu-container"
      data-dh-context-menu={true}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      role="menuitem"
      style={
        {
          "left": NaN,
          "top": NaN,
        }
      }
      tabIndex={0}
    >
      <div
        className="loading"
      >
        <div
          aria-label="Loading..."
          aria-valuemax={100}
          aria-valuemin={0}
          className="loading-spinner loading-spinner-vertical-align"
          role="progressbar"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`renders a promise returning menu items properly 1`] = `
<div>
  <div
    className="context-menu-root active"
  >
    <div
      className="context-menu-container"
      data-dh-context-menu={true}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      role="menuitem"
      style={
        {
          "left": NaN,
          "top": NaN,
        }
      }
      tabIndex={0}
    >
      <div
        className="context-menu-item"
      >
        <button
          className="btn-context-menu"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseMove={[Function]}
          title=""
          type="button"
        >
          <div
            className="btn-context-menu-wrapper"
          >
            <span
              className="icon"
            />
            <span
              className="title"
            >
              Test1
            </span>
          </div>
        </button>
      </div>
      <div
        className="context-menu-item"
      >
        <button
          className="btn-context-menu"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseMove={[Function]}
          title=""
          type="button"
        >
          <div
            className="btn-context-menu-wrapper"
          >
            <span
              className="icon"
            />
            <span
              className="title"
            >
              Test2
            </span>
          </div>
        </button>
      </div>
      <div
        className="context-menu-item"
      >
        <button
          className="btn-context-menu"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseMove={[Function]}
          title=""
          type="button"
        >
          <div
            className="btn-context-menu-wrapper"
          >
            <span
              className="icon"
            />
            <span
              className="title"
            >
              Test3
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders an empty menu for a rejected promise 1`] = `
<div>
  <div
    className="context-menu-root active"
  >
    <div
      className="context-menu-container"
      data-dh-context-menu={true}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      role="menuitem"
      style={
        {
          "left": NaN,
          "top": NaN,
        }
      }
      tabIndex={0}
    >
      <div
        className="loading"
      >
        <div
          aria-label="Loading..."
          aria-valuemax={100}
          aria-valuemin={0}
          className="loading-spinner loading-spinner-vertical-align"
          role="progressbar"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`renders without crashing 1`] = `
<div>
  <div
    className="context-menu-root"
  />
</div>
`;
